File: go/tests/rag_test.go
// rag_test.go
//
// Exercises RagContextBuilder + RagPipelineBuilder. Mirrors the TS pilot suite
// tests/rag.test.ts 1:1 plus the C# RagContextBuilderTests, the fluent-builder
// surface, and the embedder ranking path.

package circleai_test

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/google/uuid"

	circleai "github.com/bhengubv/CircleAI/go"
)

// ── Helpers ──────────────────────────────────────────────────────────────────

type ragEntryOpts struct {
	userText      string
	assistantText string
	appContext    string
	embedding     []float32
	recorded      time.Time
}

func ragEpisodic(o ragEntryOpts) circleai.EpisodicMemoryEntry {
	rec := o.recorded
	if rec.IsZero() {
		rec = time.Date(2026, 6, 1, 12, 34, 0, 0, time.UTC)
	}
	ut := o.userText
	if ut == "" {
		ut = "u"
	}
	at := o.assistantText
	if at == "" {
		at = "a"
	}
	e := circleai.EpisodicMemoryEntry{
		ID:            uuid.New(),
		RecordedAtUTC: rec,
		UserText:      ut,
		AssistantText: at,
		Embedding:     o.embedding,
	}
	if o.appContext != "" {
		ac := o.appContext
		e.AppContext = &ac
	}
	return e
}

func countOccurrences(text, token string) int {
	return strings.Count(text, token)
}

// fixedEmbedder maps any query to a fixed vector.
type fixedEmbedder struct {
	vec []float32
}

func (f fixedEmbedder) Generate(_ context.Context, _ string) ([]float32, error) {
	return f.vec, nil
}

// throwingEmbedder always fails.
type throwingEmbedder struct{}

func (throwingEmbedder) Generate(_ context.Context, _ string) ([]float32, error) {
	return nil, errors.New("embedder offline")
}

// throwingEpisodicStore always returns an error — used to test resilience.
type throwingEpisodicStore struct{}

func (throwingEpisodicStore) Add(context.Context, circleai.EpisodicMemoryEntry) error {
	return errors.New("store failure")
}
func (throwingEpisodicStore) Search(context.Context, []float32, int) ([]circleai.EpisodicMemoryEntry, error) {
	return nil, errors.New("store failure")
}
func (throwingEpisodicStore) GetRecent(context.Context, int) ([]circleai.EpisodicMemoryEntry, error) {
	return nil, errors.New("store failure")
}
func (throwingEpisodicStore) Count(context.Context) (int, error) {
	return 0, errors.New("store failure")
}
func (throwingEpisodicStore) PruneOlderThan(context.Context, time.Time) (int, error) {
	return 0, errors.New("store failure")
}

func mustBuilder(t *testing.T, store circleai.IEpisodicMemoryStore, embedder circleai.ITextEmbedder, topK, maxChars int) *circleai.RagContextBuilder {
	t.Helper()
	b, err := circleai.NewRagContextBuilder(store, embedder, topK, maxChars)
	if err != nil {
		t.Fatalf("NewRagContextBuilder: %v", err)
	}
	return b
}

// ── Constructor guards ───────────────────────────────────────────────────────

func TestRagContextBuilder_NilStore(t *testing.T) {
	if _, err := circleai.NewRagContextBuilder(nil, nil, 5, 300); err == nil {
		t.Errorf("expected error for nil store")
	}
}

// ── Empty / missing query ────────────────────────────────────────────────────

func TestRagContextBuilder_EmptyQuery(t *testing.T) {
	ctx := context.Background()
	b := mustBuilder(t, circleai.NewInMemoryEpisodicStoreDefault(), nil, 5, 300)
	if got := b.BuildContext(ctx, ""); got != "" {
		t.Errorf("empty query: got %q want empty", got)
	}
	if got := b.BuildContext(ctx, "   "); got != "" {
		t.Errorf("whitespace query: got %q want empty", got)
	}
}

// ── Empty store ──────────────────────────────────────────────────────────────

func TestRagContextBuilder_EmptyStore(t *testing.T) {
	ctx := context.Background()
	b := mustBuilder(t, circleai.NewInMemoryEpisodicStoreDefault(), nil, 5, 300)
	if got := b.BuildContext(ctx, "hello"); got != "" {
		t.Errorf("empty store: got %q want empty", got)
	}
}

// ── Formatting (recency fallback, no embedder) ───────────────────────────────

func TestRagContextBuilder_FormattedBlock(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{
		userText:      "What is SDPKT?",
		assistantText: "SDPKT is the TGN wallet.",
		recorded:      time.Date(2026, 6, 1, 11, 0, 0, 0, time.UTC),
	}))

	b := mustBuilder(t, store, nil, 3, 300)
	result := b.BuildContext(ctx, "tell me about the wallet")

	if result == "" {
		t.Fatal("result is empty")
	}
	for _, want := range []string{"What is SDPKT?", "SDPKT is the TGN wallet.", "[Relevant past exchanges"} {
		if !strings.Contains(result, want) {
			t.Errorf("result missing %q; got:\n%s", want, result)
		}
	}
}

func TestRagContextBuilder_UTCTimestampAndLabels(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{
		userText:      "q",
		assistantText: "r",
		recorded:      time.Date(2026, 6, 1, 9, 5, 0, 0, time.UTC),
	}))
	b := mustBuilder(t, store, nil, 1, 300)
	result := b.BuildContext(ctx, "anything")
	for _, want := range []string{"[2026-06-01 09:05 UTC]", "User: q", "B!: r"} {
		if !strings.Contains(result, want) {
			t.Errorf("result missing %q; got:\n%s", want, result)
		}
	}
}

func TestRagContextBuilder_RespectsTopK(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	for i := 0; i < 10; i++ {
		mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "question", assistantText: "answer"}))
	}
	b := mustBuilder(t, store, nil, 2, 300)
	result := b.BuildContext(ctx, "any question")
	if n := countOccurrences(result, "• ["); n != 2 {
		t.Errorf("bullet count: got %d want 2", n)
	}
}

func TestRagContextBuilder_IncludesAppContext(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "bid query", assistantText: "bid answer", appContext: "tgn.bidbaas"}))
	b := mustBuilder(t, store, nil, 3, 300)
	result := b.BuildContext(ctx, "bidding")
	if !strings.Contains(result, "tgn.bidbaas") {
		t.Errorf("result missing app context; got:\n%s", result)
	}
}

func TestRagContextBuilder_TruncatesLongText(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	longText := strings.Repeat("x", 500)
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: longText, assistantText: "a"}))
	// maxCharsPerEntry 100 → half 50 → truncate to 49 chars + "…"
	b := mustBuilder(t, store, nil, 1, 100)
	result := b.BuildContext(ctx, "q")
	if !strings.Contains(result, strings.Repeat("x", 49)+"…") {
		t.Errorf("expected 49 x's + ellipsis; got:\n%s", result)
	}
	if strings.Contains(result, strings.Repeat("x", 51)) {
		t.Errorf("text not truncated; got:\n%s", result)
	}
}

// ── Embedder ranking path ────────────────────────────────────────────────────

func TestRagContextBuilder_RanksByEmbedding(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "near", assistantText: "n", embedding: []float32{1, 0}}))
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "far", assistantText: "f", embedding: []float32{0, 1}}))

	// Embedder maps any query to the x-axis, so "near" should rank first.
	b := mustBuilder(t, store, fixedEmbedder{vec: []float32{1, 0}}, 1, 300)
	result := b.BuildContext(ctx, "anything")
	if !strings.Contains(result, "near") {
		t.Errorf("result missing 'near'; got:\n%s", result)
	}
	if strings.Contains(result, "far") {
		t.Errorf("result should not contain 'far'; got:\n%s", result)
	}
}

func TestRagContextBuilder_EmbedderThrowsFallsBackToRecency(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{
		userText:      "only",
		assistantText: "entry",
		recorded:      time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC),
	}))
	b := mustBuilder(t, store, throwingEmbedder{}, 3, 300)
	result := b.BuildContext(ctx, "q")
	if !strings.Contains(result, "only") {
		t.Errorf("result missing 'only'; got:\n%s", result)
	}
}

// ── Resilience ───────────────────────────────────────────────────────────────

func TestRagContextBuilder_StoreThrowsReturnsEmpty(t *testing.T) {
	ctx := context.Background()
	b := mustBuilder(t, throwingEpisodicStore{}, nil, 5, 300)
	if got := b.BuildContext(ctx, "query"); got != "" {
		t.Errorf("store throws: got %q want empty", got)
	}
}

// ── RagPipelineBuilder ───────────────────────────────────────────────────────

func TestRagPipelineBuilder_FromStore(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "hi", assistantText: "hello"}))
	rag, err := circleai.NewRagPipelineBuilder().WithStore(store).WithTopK(2).WithMaxCharsPerEntry(500).Build()
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if ctxStr := rag.BuildContext(ctx, "greeting"); !strings.Contains(ctxStr, "hi") {
		t.Errorf("result missing 'hi'; got:\n%s", ctxStr)
	}
}

func TestRagPipelineBuilder_WithInMemoryStore(t *testing.T) {
	ctx := context.Background()
	rag, err := circleai.NewRagPipelineBuilder().WithInMemoryStore().Build()
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if got := rag.BuildContext(ctx, "nothing stored"); got != "" {
		t.Errorf("got %q want empty", got)
	}
}

func TestRagPipelineBuilder_BuildWithoutStoreThrows(t *testing.T) {
	if _, err := circleai.NewRagPipelineBuilder().Build(); err == nil {
		t.Errorf("expected error building without a store")
	}
}

func TestRagPipelineBuilder_TopKRejectsBelowOne(t *testing.T) {
	if _, err := circleai.NewRagPipelineBuilder().WithInMemoryStore().WithTopK(0).Build(); err == nil {
		t.Errorf("expected error for topK=0")
	}
}

func TestRagPipelineBuilder_MaxCharsRejectsBelow50(t *testing.T) {
	if _, err := circleai.NewRagPipelineBuilder().WithInMemoryStore().WithMaxCharsPerEntry(49).Build(); err == nil {
		t.Errorf("expected error for maxChars=49")
	}
}

func TestRagPipelineBuilder_WithEmbedder(t *testing.T) {
	ctx := context.Background()
	store := circleai.NewInMemoryEpisodicStoreDefault()
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "near", assistantText: "n", embedding: []float32{1, 0}}))
	mustAdd(t, store, ragEpisodic(ragEntryOpts{userText: "far", assistantText: "f", embedding: []float32{0, 1}}))
	rag, err := circleai.NewRagPipelineBuilder().WithStore(store).WithEmbedder(fixedEmbedder{vec: []float32{1, 0}}).WithTopK(1).Build()
	if err != nil {
		t.Fatalf("Build: %v", err)
	}
	if ctxStr := rag.BuildContext(ctx, "q"); !strings.Contains(ctxStr, "near") {
		t.Errorf("result missing 'near'; got:\n%s", ctxStr)
	}
}

File: rust/tests/rag_test.rs
//! rag_test.rs
//!
//! Exercises RagContextBuilder + RagPipelineBuilder. Mirrors the TS suite
//! tests/rag.test.ts and CircleAI.Tests.RagContextBuilderTests plus the fluent
//! builder surface and the embedder ranking path.

use std::sync::Arc;

use chrono::{TimeZone, Utc};
use circle_ai::brain::BrainError;
use circle_ai::memory::episodic::InMemoryEpisodicStore;
use circle_ai::memory::rag::{
    ITextEmbedder, RagContextBuilder, RagEpisodicStore, RagPipelineBuilder,
};
use circle_ai::memory::EpisodicMemoryEntry;
use uuid::Uuid;

// ── Helpers ──────────────────────────────────────────────────────────────────

struct EpisodicOpts<'a> {
    user_text: &'a str,
    assistant_text: &'a str,
    app_context: Option<&'a str>,
    embedding: Option<Vec<f32>>,
    recorded: Option<chrono::DateTime<Utc>>,
}

impl Default for EpisodicOpts<'_> {
    fn default() -> Self {
        Self {
            user_text: "u",
            assistant_text: "a",
            app_context: None,
            embedding: None,
            recorded: None,
        }
    }
}

fn episodic(o: EpisodicOpts) -> EpisodicMemoryEntry {
    EpisodicMemoryEntry {
        id: Uuid::new_v4(),
        recorded_at_utc: o
            .recorded
            .unwrap_or_else(|| Utc.with_ymd_and_hms(2026, 6, 1, 12, 34, 0).unwrap()),
        user_text: o.user_text.to_string(),
        assistant_text: o.assistant_text.to_string(),
        app_context: o.app_context.map(|s| s.to_string()),
        embedding: o.embedding,
        tags: None,
    }
}

fn count_occurrences(text: &str, token: &str) -> usize {
    text.matches(token).count()
}

/// Store that always fails — used to test resilience (mirrors the TS
/// ThrowingEpisodicStore).
struct ThrowingEpisodicStore;
impl RagEpisodicStore for ThrowingEpisodicStore {
    fn search(
        &self,
        _query_embedding: Option<&[f32]>,
        _top_k: usize,
    ) -> Result<Vec<EpisodicMemoryEntry>, BrainError> {
        Err(BrainError::new("store failure"))
    }
}

/// Embedder that maps any query to a fixed vector.
struct FixedEmbedder(Vec<f32>);
impl ITextEmbedder for FixedEmbedder {
    fn generate(&self, _text: &str) -> Result<Vec<f32>, BrainError> {
        Ok(self.0.clone())
    }
}

/// Embedder that always fails.
struct FailingEmbedder;
impl ITextEmbedder for FailingEmbedder {
    fn generate(&self, _text: &str) -> Result<Vec<f32>, BrainError> {
        Err(BrainError::new("embedder offline"))
    }
}

fn add(store: &InMemoryEpisodicStore, e: EpisodicMemoryEntry) {
    store.add_shared(e).unwrap();
}

// ══════════════════════════════════════════════════════════════════════════
// Empty / missing query
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn empty_query_returns_empty() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    let b = RagContextBuilder::with_store(store);
    assert_eq!(b.build_context(""), "");
}

#[test]
fn whitespace_query_returns_empty() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    let b = RagContextBuilder::with_store(store);
    assert_eq!(b.build_context("   "), "");
}

// ══════════════════════════════════════════════════════════════════════════
// Empty store
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn empty_store_returns_empty() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    let b = RagContextBuilder::with_store(store);
    assert_eq!(b.build_context("hello"), "");
}

// ══════════════════════════════════════════════════════════════════════════
// Non-empty store — recency fallback (no embedder)
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn returns_a_formatted_block_with_the_header_and_both_texts() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "What is SDPKT?",
            assistant_text: "SDPKT is the TGN wallet.",
            recorded: Some(Utc.with_ymd_and_hms(2026, 6, 1, 11, 0, 0).unwrap()),
            ..Default::default()
        }),
    );

    let b = RagContextBuilder::new(store, None, 3, 300);
    let result = b.build_context("tell me about the wallet");

    assert_ne!(result, "");
    assert!(result.contains("What is SDPKT?"));
    assert!(result.contains("SDPKT is the TGN wallet."));
    assert!(result.contains("[Relevant past exchanges"));
}

#[test]
fn formats_the_utc_timestamp_and_labels_user_and_b() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "q",
            assistant_text: "r",
            recorded: Some(Utc.with_ymd_and_hms(2026, 6, 1, 9, 5, 0).unwrap()),
            ..Default::default()
        }),
    );
    let b = RagContextBuilder::new(store, None, 1, 300);
    let result = b.build_context("anything");
    assert!(result.contains("[2026-06-01 09:05 UTC]"));
    assert!(result.contains("User: q"));
    assert!(result.contains("B!: r"));
}

#[test]
fn respects_top_k() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    for i in 0..10 {
        add(
            &store,
            episodic(EpisodicOpts {
                user_text: &format!("question {i}"),
                assistant_text: &format!("answer {i}"),
                ..Default::default()
            }),
        );
    }

    let b = RagContextBuilder::new(store, None, 2, 300);
    let result = b.build_context("any question");
    assert_eq!(count_occurrences(&result, "• ["), 2);
}

#[test]
fn includes_the_app_context_when_set() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "bid query",
            assistant_text: "bid answer",
            app_context: Some("tgn.bidbaas"),
            ..Default::default()
        }),
    );
    let b = RagContextBuilder::new(store, None, 3, 300);
    let result = b.build_context("bidding");
    assert!(result.contains("tgn.bidbaas"));
}

#[test]
fn truncates_long_texts_to_half_the_budget_with_an_ellipsis() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    let long_text = "x".repeat(500);
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: &long_text,
            assistant_text: "a",
            ..Default::default()
        }),
    );
    // max_chars_per_entry 100 → half 50 → truncate to 49 chars + "…"
    let b = RagContextBuilder::new(store, None, 1, 100);
    let result = b.build_context("q");
    assert!(result.contains(&(("x".repeat(49)) + "…")));
    assert!(!result.contains(&"x".repeat(51)));
}

// ══════════════════════════════════════════════════════════════════════════
// Embedder ranking path
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn ranks_by_the_embedding_when_an_embedder_is_supplied() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "near",
            assistant_text: "n",
            embedding: Some(vec![1.0, 0.0]),
            ..Default::default()
        }),
    );
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "far",
            assistant_text: "f",
            embedding: Some(vec![0.0, 1.0]),
            ..Default::default()
        }),
    );

    // Embedder maps any query to the x-axis, so "near" should rank first.
    let embedder: Arc<dyn ITextEmbedder> = Arc::new(FixedEmbedder(vec![1.0, 0.0]));
    let b = RagContextBuilder::new(store, Some(embedder), 1, 300);
    let result = b.build_context("anything");
    assert!(result.contains("near"));
    assert!(!result.contains("far"));
}

#[test]
fn falls_back_to_recency_when_the_embedder_fails() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "only",
            assistant_text: "entry",
            recorded: Some(Utc.with_ymd_and_hms(2026, 6, 1, 0, 0, 0).unwrap()),
            ..Default::default()
        }),
    );
    let embedder: Arc<dyn ITextEmbedder> = Arc::new(FailingEmbedder);
    let b = RagContextBuilder::new(store, Some(embedder), 3, 300);
    let result = b.build_context("q");
    assert!(result.contains("only"));
}

// ══════════════════════════════════════════════════════════════════════════
// Resilience — store fails
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn returns_empty_when_the_store_fails() {
    let store: Arc<dyn RagEpisodicStore> = Arc::new(ThrowingEpisodicStore);
    let b = RagContextBuilder::with_store(store);
    assert_eq!(b.build_context("query"), "");
}

// ══════════════════════════════════════════════════════════════════════════
// RagPipelineBuilder
// ══════════════════════════════════════════════════════════════════════════

#[test]
fn builds_from_an_in_memory_store_and_produces_a_working_builder() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "hi",
            assistant_text: "hello",
            ..Default::default()
        }),
    );
    let rag = RagPipelineBuilder::create()
        .with_store(store)
        .with_top_k(2)
        .unwrap()
        .with_max_chars_per_entry(500)
        .unwrap()
        .build()
        .unwrap();
    let ctx = rag.build_context("greeting");
    assert!(ctx.contains("hi"));
}

#[test]
fn with_in_memory_store_wires_a_fresh_store() {
    let rag = RagPipelineBuilder::create()
        .with_in_memory_store()
        .build()
        .unwrap();
    assert_eq!(rag.build_context("nothing stored"), "");
}

#[test]
fn build_without_a_store_fails() {
    // Map to the error string first — RagContextBuilder is not Debug, so
    // unwrap_err() would not compile.
    let msg = RagPipelineBuilder::create()
        .build()
        .err()
        .map(|e| e.to_string());
    assert!(msg
        .as_deref()
        .unwrap_or("")
        .to_lowercase()
        .contains("episodic memory store is required"));
}

#[test]
fn with_top_k_rejects_values_below_1() {
    assert!(RagPipelineBuilder::create().with_top_k(0).is_err());
}

#[test]
fn with_max_chars_per_entry_rejects_values_below_50() {
    assert!(RagPipelineBuilder::create()
        .with_max_chars_per_entry(49)
        .is_err());
}

#[test]
fn with_embedder_wires_the_semantic_ranking_seam() {
    let store = Arc::new(InMemoryEpisodicStore::with_default_capacity());
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "near",
            assistant_text: "n",
            embedding: Some(vec![1.0, 0.0]),
            ..Default::default()
        }),
    );
    add(
        &store,
        episodic(EpisodicOpts {
            user_text: "far",
            assistant_text: "f",
            embedding: Some(vec![0.0, 1.0]),
            ..Default::default()
        }),
    );
    let embedder: Arc<dyn ITextEmbedder> = Arc::new(FixedEmbedder(vec![1.0, 0.0]));
    let rag = RagPipelineBuilder::create()
        .with_store(store)
        .with_embedder(embedder)
        .with_top_k(1)
        .unwrap()
        .build()
        .unwrap();
    let ctx = rag.build_context("q");
    assert!(ctx.contains("near"));
}
